File: epi/KthLargestInArray.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.Collections;
import java.util.List;
import java.util.Random;

public class KthLargestInArray {
  // The numbering starts from one, i.e., if A = [3,1,-1,2] then
  // findKthLargest(1, A) returns 3, findKthLargest(2, A) returns 2,
  // findKthLargest(3, A) returns 1, and findKthLargest(4, A) returns -1.
  @EpiTest(testDataFile = "kth_largest_in_array.tsv")
  public static int findKthLargest(int k, List<Integer> A) {
//    Collections.sort(A, Collections.reverseOrder());
//    return A.get(k - 1);
    //Quick Select
    /*
      (3, 2,1,5,4)
      k 1
     */
    Integer[] nums = new Integer[A.size()];
    for(int i = 0; i < nums.length; i++) {
      nums[i] = A.get(i);
    }
    int l = 0, r = nums.length - 1, m = 0;
    k--;
    k = nums.length - 1 - k;
    while(l <= r) {
      m = new Random().nextInt(r - l + 1) + l;
      swap(nums, m, r);
      int pivot = pagination(nums, l, r);
      if(k == pivot) {
        return nums[pivot].intValue();
      } else if(k < pivot) {
        r = pivot - 1;
      } else{
        l = pivot + 1;
      }
    }
    return -1;
  }

  private static int pagination(Integer[] nums, int l, int r) {
    int idx = l - 1;
    int target = nums[r];

    for(int i = l; i < r; i++) {
      if(nums[i] < target) {
        idx++;
        swap(nums, i, idx);
      }
    }

    swap(nums, r, idx + 1);
    return idx + 1;
  }

  private static void swap(Integer[] nums, int i, int j) {
    Integer tmp = nums[i];
    nums[i] = nums[j];
    nums[j] = tmp;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "KthLargestInArray.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/KthLargestInArray.kt
package epi.kt

import java.lang.RuntimeException
import java.util.*

object KthLargestInArray {
    fun findKthLargestStable(k : Int, A : MutableList<Int>) : Int {
        //Use merge sort directly
        mergeSort(A, 0, A.size - 1)
        return A[A.size - k]
    }

    private fun mergeSort(A : MutableList<Int>, l : Int, r : Int) {
        if(l < r) {
            val m = l + (r - l) / 2
            mergeSort(A, l, m)
            mergeSort(A, m + 1, r)
            merge(A, l, m , r)
        }
    }

    private fun merge(A: MutableList<Int>, l : Int, m : Int, r : Int) {
        val left = ArrayList(A.subList(l, m + 1))
        val right = ArrayList(A.subList(m + 1, r + 1))

        var i = l//Extreme important!!
        var lI = 0
        var rI = 0

        while(lI < left.size && rI < right.size) {
            if(left[lI] < right[rI]) {
                A[i++] = left[lI++]
            } else {
                A[i++] = right[rI++]
            }
        }

        while(rI < right.size) {
            A[i++] = right[rI++]
        }

        while(lI < left.size) {
            A[i++] = left[lI++]
        }
    }

    fun findMedian(A : List<Int>) : Int {
        if(A.size % 2 == 0) {
            //med = A[A.size / 2] + A[A.size / 2 - 1] / 2
        }

        throw RuntimeException("Not impl!")
    }
    /*
        3 2 1 5 4
        3
     */
    fun findKthLargest(k : Int, A : List<Int>) : Int {
        //Return val
        val list: MutableList<Int> = ArrayList()
        //T O(A) S O(A)
        for(i in A.indices) {
            list.add(A[i])
        }

        var pivot = -1
        var st = 0
        var ed = list.size - 1
        var lK = k - 1
        //T O(logA) S O(A)
        while(true) {
            val ranIdx = Random().nextInt(ed - st + 1) + st
            Collections.swap(list, ranIdx, ed)
            pivot = partition(list, st, ed)
            if(pivot == lK) {
                break
            } else if(pivot < lK) {
                st = pivot + 1
            } else {
                ed = pivot - 1
            }
        }

        return list[pivot]
    }

    private fun partition(list : MutableList<Int>, st : Int, ed : Int) : Int {
        val target = list[ed]
        var idx = st - 1

        for(i in st until ed) {
            if(list[i] > target) {
                Collections.swap(list, ++idx, i)
            }
        }

        Collections.swap(list, idx + 1, ed)
        return idx + 1
    }
}